python/lsst/daf/butler/registry/collections/synthIntKey.py

```diff
@@ -0,0 +1,133 @@
+    table schema.
+    """
+
+    @classmethod
+    def initialize(cls, db: Database, context: StaticTablesContext) -> SynthIntKeyCollectionManager:
+        # Docstring inherited from CollectionManager.
+        return cls(db, tables=context.addTableTuple(_TABLES_SPEC),
+                   collectionIdName="collection_id")
+
+    @classmethod
+    def addCollectionForeignKey(cls, tableSpec: ddl.TableSpec, *, prefix: str = "collection",
+                                onDelete: Optional[str] = None, **kwds: Any) -> ddl.FieldSpec:
+        # Docstring inherited from CollectionManager.
+        if prefix is None:
+            prefix = "collection"
+        original = _TABLES_SPEC.collection.fields["collection_id"]
+        copy = ddl.FieldSpec(cls.getCollectionForeignKeyName(prefix), dtype=original.dtype, **kwds)
+        tableSpec.fields.add(copy)
+        tableSpec.foreignKeys.append(ddl.ForeignKeySpec("collection", source=(copy.name,),
+                                                        target=(original.name,), onDelete=onDelete))
+        return copy
+
+    @classmethod
+    def addRunForeignKey(cls, tableSpec: ddl.TableSpec, *, prefix: str = "run",
+                         onDelete: Optional[str] = None, **kwds: Any) -> ddl.FieldSpec:
+        # Docstring inherited from CollectionManager.
+        if prefix is None:
+            prefix = "run"
+        original = _TABLES_SPEC.run.fields["collection_id"]
+        copy = ddl.FieldSpec(cls.getRunForeignKeyName(prefix), dtype=original.dtype, **kwds)
+        tableSpec.fields.add(copy)
+        tableSpec.foreignKeys.append(ddl.ForeignKeySpec("run", source=(copy.name,),
+                                                        target=(original.name,), onDelete=onDelete))
+        return copy
+
+    def _setRecordCache(self, records: Iterator[CollectionRecord]):
+        """Set internal record cache to contain given records,
+        old cached records will be removed.
+        """
+        self._records = {}
+        self._nameCache = {}
+        for record in records:
+            self._records[record.key] = record
+            self._nameCache[record.name] = record
+
+    def _addCachedRecord(self, record: CollectionRecord):
+        """Add single record to cache.
+        """
+        if not self._nameCache:
+            self._nameCache = {}
+        self._records[record.key] = record
+        self._nameCache[record.name] = record
+
+    def _removeCachedRecord(self, record: CollectionRecord):
+        """Remove single record from cache.
+        """
+        del self._records[record.key]
+        if self._nameCache:
+            del self._nameCache[record.name]
+
+    def _getByName(self, name: str):
+        # Docstring inherited from DefaultCollectionManager.
+        if self._nameCache:
+            return self._nameCache.get(name)
+
+    _nameCache: Optional[Mapping[str, CollectionRecord]] = None
+    """Mapping that indexes records by name for fast search.
+    """
```

Comment:
Elsewhere in daf_butler I think we've only been documenting/annotating public instance attributes like this.  I think private ones probably shouldn't have docs, and I learned today that putting the type annotations in `__init__` is sufficient for mypy.  This may be something we should discuss more broadly before assuming precedent -> policy, but there is precedent.

Reply:
I really wanted to have this member defined in `__init__` but that is a bit more code and documentation, this is why I used this shortcut of defining a private attribute as a class member and felt that as a class member it should be documented. Proabably I just need to add `__init__` and not expose this at class level.